packages/opentelemetry-core/src/common/attributes.ts

```diff
@@ -47,6 +47,65 @@ export function isAttributeValue(val: unknown): val is AttributeValue {
   return isValidPrimitiveAttributeValue(val);
 }
 
+export function truncateValueIfTooLong(
+  value: AttributeValue,
+  limit: number | undefined,
+  logger: Logger
+): AttributeValue {
+  if (limit == null) {
+    return value;
+  }
+
+  if (typeof value === 'boolean' || typeof value === 'number') {
+    // these types reasonably can't exceed a limit
+    // if the limit was set to e.g. 2, then true with a length of 4 will exceed a limit
+    // however, the purpose of the limit is to prevent accidents such as values being thousands of characters long
+    return value;
+  }
+
+  if (Array.isArray(value)) {
+    let accruedLength = 0;
+
+    // roughly measure the size of a serialized array
+    // this is done to avoid running json stringify each time attributes are added
+    for (const element of value) {
+      if (element == undefined) {
+        accruedLength += 4;
+      } else if (typeof element === 'string') {
+        accruedLength += element.length;
+      } else {
+        accruedLength += element.toString().length;
+      }
+
+      if (accruedLength > limit) {
+        break;
+      }
+    }
+
+    if (accruedLength > limit) {
```

Comment:
not sure why not stringify it first and skip all that counting. Sooner or later this value will be stringified anyway (either now or later in exporter) so counting value and stringify seems like extra overhead.

Reply:
Arrays truncation is now based on `JSON.stringify` and backed by spec changes: "SDKs SHOULD choose any serialization protocol, which is performant and appropriate for the language and/or environment they are implemented in."